src/stories/components/UploadPhoto/UploadPhoto.js

```diff
@@ -1,118 +1,181 @@
 import './styles.scss';
 
-import React, { useState } from 'react';
+import React, { useState, useEffect } from 'react';
+import { useParams } from 'react-router-dom';
 import Webcam from 'react-webcam';
 
 import { PhotoInstructions } from '../PhotoInstructions/PhotoInstructions';
 import { Button } from '../Button/Button';
 import { ImageGuidelines } from '../ImageGuidelines/ImageGuidelines';
 import { ToggleCamera } from '../ToggleCamera/ToggleCamera';
 import { FileUpload } from '../FileUpload/FileUpload';
-import { uploadSelfImage }
+import { uploadSelfImage, observeStimuliCompletion }
   from '../../../firebase/api/gcp-utils';
 import { Loader } from '../Loader/Loader';
 import { StatusCodes } from 'http-status-codes';
 
+// Milliseconds to wait before attempting to fetch generated stimuli URLs
+const STIMULI_GENERATION_WAIT_1 = 40000;
+const STIMULI_GENERATION_WAIT_2 = 10000;
+
 /**
- * Component for webcam controls element.
+ * Component for webcam controls and photo uploading.
  *
  * @component
  * @return {object} (
  *   <UploadPhoto />
  * )
  */
 export const UploadPhoto = () => {
+  // ===== STATE ==============================================================
+  const { experimentId, participantId } = useParams(); // Parse URL params
   const [cameraIsOn, setWebcamOn] = useState(false);
-  const [file, setFile] = useState('');
-  const [image, setImage] = useState('');
+  const [file, setFile] = useState(''); // For photos uploaded through form
+  const [image, setImage] = useState(''); // For photos taken with webcam
   const [error, setError] = useState(false);
   const webcamRef = React.useRef(null);
   const [loading, setLoading] = useState(false);
   const [complete, setComplete] = useState(false);
-  // TODO: define urls state
+  const [stimuliUrls, setStimuliUrls] = useState([]);
+
+  // ===== STATUS CHECKS ======================================================
+  // Check for stimuli generation completion; 1st try.
+  useEffect(() => {
+    if (loading) {
+      // Wait 40 seconds for the stimuli to be generated
+      setTimeout(() => {
+        observeStimuliCompletion(
+          participantId,
+          experimentId,
+          setStimuliUrls,
+          stimuliFetchingFirstTryFailed,
+        );
+      }, STIMULI_GENERATION_WAIT_1);
+    }
+  }, [loading]);
+
+  // Check if stimuli URLs have been fetched
+  useEffect(() => {
+    if (stimuliUrls.length > 0) {
+      checkStimuli();
```

Comment:
This is a little weird.

Looking at checkStimuli(), we also do a check for stimuliUrls.length > 0 there as well?